lumen/ai/llm.py

```diff
@@ -168,38 +171,40 @@ class OpenAI(Llm):
 
     base_url = param.String()
 
-    model_name = param.String()
-
     mode = param.Selector(default=Mode.FUNCTIONS)
 
     temperature = param.Number(default=0.2, bounds=(0, None), constant=True)
 
     organization = param.String()
 
-    _models = {
-        "gpt-3.5-turbo": {"model_name": "gpt-3.5-turbo"},
-        "gpt-4": {"model_name": "gpt-4"},
-        "gpt-4-turbo-preview": {"model_name": "gpt-4-turbo-preview"},
+    _models_kwargs = {
+        "default": {"model": "gpt-3.5-turbo"},
+        "reasoning": {"model": "gpt-4-turbo-preview"},
     }
 
     @property
     def _client_kwargs(self):
-        return {"model": self.model_name, "temperature": self.temperature}
+        return {"temperature": self.temperature}
+
+    def get_client(self, model_key: str):
+        if client := pn.state.cache.get(model_key):
+            return client
 
-    def _init_model(self):
         import openai
 
-        model_kwargs = {}
+        model_kwargs = self._get_model_kwargs(model_key)
+        model = model_kwargs.pop("model", "default")
         if self.base_url:
             model_kwargs["base_url"] = self.base_url
         if self.api_key:
             model_kwargs["api_key"] = self.api_key
         if self.organization:
             model_kwargs["organization"] = self.organization
-        self._model = openai.AsyncOpenAI(**model_kwargs)
-        self._raw_client = self._model.chat.completions.create
-        self._client = self._create_client(self._raw_client)
-
+        llm = openai.AsyncOpenAI(**model_kwargs)
+        raw_client = llm.chat.completions.create
+        client = self._create_client(raw_client, model)
+        pn.state.cache[model_key] = client
```

Comment:
Why cache?

Reply:
For consistency across the `get_client` methods, but I guess compared to LlamaCpp, it's not really necessary to cache.

Reply:
Indeed, seems pointless here.